Design/Workbench/design-projectsaving.md

```diff
@@ -0,0 +1,229 @@
+﻿
+Project Saving for Mantid Workbench
+==============================
+Motivation
+----------
+In MantidPlot it was possible to save all currently open windows and all workspaces, and then be able to load then back in the exact(-ish) state that they were in previously. To achieve this goal again in Mantid Workbench, the problem will be split into two where window serialization and saving and therefore reopening will be separate in implementation, from saving workspaces.
+
+The reason for the split is so that it can be handled in a more abstract manor. Instead of thinking of the task as one great big issue it can be split into the two smaller problems and then split further from there, later on in design.
+
+This document exists because Project Saving is a non-trivial challenge to tackle when it comes to ideas with saving and loading interfaces created in both Python and C++ as well as saving and loading of workspaces. This issue should be fairly simple to understand, simply users should be able to save their projects and load them back in, to at least the standard that MantidPlot was able to achieve.
+
+
+Functional Requirements
+-----------------------
+- Graceful handling of other versions that have been saved previously
+- Support for both Python and C++ interfaces
+- Portable across OS's
+  - However gracefully handle missing files such as scripts or removed workspaces etc
+- Independant File reading and writing implementation from the rest of this section of code so it can be removed and replaced without changing much else
+- Ability to partially save a project as well as save it's entirity
+- Be able to save and load, plots, interfaces, workspaces, workspaces interfaces and scripts.
+  - Scripts only lightly and not transferrable across machines/operating systems
+
+Changes to MantidWorkbench needed
+---------------------------------
+- MainWindow should track all of the open windows (as well as if they are serializable or not) similar to MantidPlot
+- File->Save Project needs to be added
+- File->Load Project needs to be added
+- A GUI for picking where to Save
+
+GUI and integration initial ideas
+---------------------------------
+It is true that this document is mostly focused on the actual implementation of saving the project and they intricacies of how that would work. However, it is important that this is discussed to some extent before hand. The intial thought process is that you would be able to do both a basic SaveAll and Save Advanced. SaveAll will run through a pretty simple Save everything as long as it's not overwriting workspaces. Save Advanced would allow the users to Select workspaces and windows individually, also adding an option as to whether or not they want to overwrite the workspaces that may or may be saved which is so at worst case users can save everything from scratch.
+
+Options for Implementation of Workspace Saving
```

Comment:
Could we remove this entirely. I think offering the possibility of changing the workspace save mechanism is well outside of the scope here and this slightly suggests that it's worth thinking about. Maybe just have this section say we will use the existing workspace saving mechansims?

Reply:
Removed the vast majority of it but I will leave a bit to explain what to do with workspaces.